src/cached_object_store.rs

```diff
@@ -114,50 +257,51 @@ impl CachedObjectStore {
         })
     }
 
-    // TODO: implement the put with cache here
-    #[allow(unused)]
-    async fn cached_put_opts(
+    // Disk cache miss, increase the ghost queue capacity leads to higher replacement rate
+    // and cached blocks will be removed faster to make room for popular blocks
+    fn increase_ghost_capacity(
         &self,
-        location: &Path,
-        payload: object_store::PutPayload,
-        opts: object_store::PutOptions,
-    ) -> object_store::Result<PutResult> {
-        self.object_store.put_opts(location, payload, opts).await
+        ghost_guard: &mut MutexGuard<'_, GhostQueue<Path, ObjectMeta>>,
+    ) {
+        let old_capacity = ghost_guard.capacity().get();
+        let new_capacity =
+            self.max_cache_size_bytes
+                .map_or((old_capacity + 3).min(64), |max_cache_size_bytes| {
+                    ((max_cache_size_bytes as f64 * 0.9).floor() as usize).min(
+                        old_capacity
+                            + (max_cache_size_bytes as f64 / old_capacity as f64).floor() as usize,
+                    )
+                });
+
+        ghost_guard
+            .adjust_capacity(NonZeroUsize::new(new_capacity).expect(
+                "LRU capacity should not be 0 and this situation should not happen at all",
+            ));
     }
 
-    // if an object is not cached before, maybe_prefetch_range will try to prefetch the object from the
-    // object store and save the parts into the local disk cache. the prefetching is helpful to reduce the
-    // number of GET requests to the object store, it'll try to aggregate the parts among the range into a
-    // single GET request, and save the related parts into local disks together.
-    // when it sends GET requests to the object store, the range is expected to be ALIGNED with the part
-    // size.
-    async fn maybe_prefetch_range(
+    // Disk cache hit, decrease the ghost queue capacity to reduce replacement rate and keep popular
+    // objects in the cache for a longer period of time
+    fn decrease_ghost_capacity(
```

Comment:
could this function placed inside `GhostQueue`? if this capacity is refreshed when new items is accessed from `GhostQueue`'s perspective, can it be placed inside `record()` or `track_accessed()` function?

Reply:
I can try it.

Reply:
I've thought about this carefully, and in my design the ghost queue should not be aware of `max_cache_size_bytes`, which would allow for decoupling. 

If we move the logic of adjusting the capacity inside the ghost queue, it means that the ghost queue needs to be aware of the existence of the `CacheObjectStore` (by passing `max_cache_size_bytes` as an argument), and if we pass the computed capacity as an argument, it's the same as the current implementation.

Reply:
In addition, whenever we miss the local disk cache, we adjust the capacity of the ghost queue, so we can't just put it in the `record`.
![LARC](https://github.com/user-attachments/assets/b6560902-b1cc-499f-8823-29f246139763)

Reply:
i suppose passing `max_cache_size_bytes` as an argument is not a coupling? 🤔 comparing to passing this argument, i feel letting the caller maintaining ghost queue's status is a bigger coupling.

(i'd still prefer a dumb `max_ghost_queue_size` config through...)

Reply:
> we miss the local disk cache, we adjust the capacity of the ghost queue

what about adding a `track_missed` in the ghost queue's interface?

Reply:
Let me think about it tomorrow.

Reply:
https://github.com/slatedb/slatedb/pull/284/files#diff-21762eb4f927def3c995cbe9d3ac7c2243c414ada7a0f57b245e1d314947883eR108
After thinking about it, I think the current implementation is clear enough. Considering the ghost capacity reduce situation, we don’t have any `record` calls.

Even if we want to modify it, it is just to let the ghost queue maintain a `max_cache_size_bytes` field, and then call something like `ghost_guard.decrease_capacity`. I think this loses the versatility of the current ghost queue.

Then considering the local cache miss scenario, we need to increase the capacity. If we use an interface such as `track_missed`, we need to determine whether the current `Path` needs to be cached in the local cache.We have two options: one is to return a bool for `track_missed`, so the logic are not as clear as the current one, or let the ghost queue hold a reference to `CacheObjectStore`, so the semantics are not clear enough and lead to a tight coupling of the two logics.

I think the current implementation maintains the generic of ghost queue (can be used as a general cache), and also allows new contributors to understand what is happening here when reading the code of `cache_object_store.rs`

Reply:
> The other nice thing about this is that we can easily enable/disable the ghost queue by just supplying a no-op impl that always returns true

Just caught up on the discord discussion. IIUC this is basically what @flaneur2020 suggested there too 😄